src/ape_cache/query.py

```diff
@@ -0,0 +1,338 @@
+        return text(
+            """
+    SELECT :columns
+    FROM contract_events
+    WHERE contract_events.block_number >= :start_block
+    AND contract_events.block_number <= :stop_block
+    AND contract_events.block_number % :step = 0
+        """
+        ).bindparams(
+            columns=",".join(query.columns),
+            start_block=query.start_block,
+            stop_block=query.stop_block,
+            step=query.step,
+        )
+
+    def perform_query(self, query: QueryType) -> Optional[pd.DataFrame]:  # type: ignore
+        try:
+            with self.database_connection as conn:
+                result = conn.execute(self.perform_query_clause(query))
+
+                if not result:
+                    # NOTE: Should be unreachable if estimated correctly
+                    raise QueryEngineError(f"Could not perform query:\n{query}")
+
+                # TODO: Fix this, should return an iterator
+                return pd.DataFrame(columns=query.columns, data=result.fetchall())
+
+        except QueryEngineError as err:
+            logger.error(f"Database not initiated: {str(err)}")
+
+    @singledispatchmethod
+    def cache_update_clause(self, query: QueryType) -> Optional[Insert]:
+        pass  # Can't cache this query
+
+    @cache_update_clause.register
+    def cache_update_block_clause(self, query: BlockQuery) -> Optional[Insert]:
+        return insert(Blocks)  # type: ignore
+
+    @cache_update_clause.register
+    def cache_update_block_txns_clause(self, query: BlockTransactionQuery) -> Optional[Insert]:
+        return insert(Transactions)  # type: ignore
+
+    @cache_update_clause.register
+    def cache_update_events_clause(self, query: ContractEventQuery) -> Optional[Insert]:
+        return insert(ContractEvents)  # type: ignore
+
+    @singledispatchmethod
+    def get_cache_data(
+        self, query: QueryType, result: Iterator[BaseInterfaceModel]
+    ) -> Optional[List[Dict[str, Any]]]:
+        raise QueryEngineError(
+            """
+            Not a compatible QueryType. For more details see our docs
+            https://docs.apeworx.io/ape/stable/methoddocs/exceptions.html#ape.exceptions.QueryEngineError
+            """
+        )
+
+    @get_cache_data.register
+    def get_block_cache_data(
+        self, query: BlockQuery, result: Iterator[BaseInterfaceModel]
+    ) -> Optional[List[Dict[str, Any]]]:
+        return [m.dict(by_alias=False) for m in result]
+
+    @get_cache_data.register
+    def get_block_txns_data(
+        self, query: BlockTransactionQuery, result: Iterator[BaseInterfaceModel]
+    ) -> Optional[List[Dict[str, Any]]]:
+        new_result = []
+        for val in [m for m in result]:
+            new_dict = {
+                k: v
+                for k, v in val.dict(by_alias=False).items()
+                if k in [c.key for c in Transactions.__table__.columns]  # type: ignore
+            }
+            new_dict["txn_hash"] = val.txn_hash
```

Comment:
This is a property from `TransactionAPI`, but mypy complains that this attribute does not exist. How do we work around this?

Reply:
I am currently removing elements with an empty receiver

Reply:
Should figure out why this is the case